### ISSUE TITLE
Add a lint for items used before definition

### ISSUE THREAD

**Issue by rodrimati1992** [rust-lang/rust-clippy#5352]
I would like to have a lint like the `items_after_statements` lint,but which only warns if an item defined inside an expression block is used before it's defined.

For example,it would trigger with this code:
```rust
fn hello() {
    println!("First statement");
    let hello: Bar = (0..10).collect();
    type Bar = Vec<u32>;
}
```

But it would **not** trigger with this code:
```rust
fn hello() {
    println!("First statement");
    type Bar = Vec<u32>;
    let hello: Bar = (0..10).collect();
}
```

**Comment by DevAccentor** [comment]
The lint already exists and does that. The issue can be closed.

**Comment by Rqnsom** [comment]
Can confirm. This is covered by `clippy::items-after-statements`. We can close this issue

**Comment by xFrednet** [comment]
So, this lint proposal is basically just a lighter version of `clippy::items-after-statements`. Adding a new lint for this SGTM. Though I would suggest a different name, since `clippy::declare-items-before-usage` doesn't work too well with lint attributes: `#[deny(clippy::declare-items-before-usage)]` does exactly the opposite of what it says. `clippy::items-before-usage` would be good IMO. Also, because it's similar to the other lint :upside_down_face:

**Comment by flip1995** [comment]
I would extend this lint with a configuration option, to only warn if the item is used before. This configuration should be turned off by default to keep the current behavior.
